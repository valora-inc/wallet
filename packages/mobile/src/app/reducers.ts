import { Platform } from 'react-native'
import { Actions, ActionTypes, AppState } from 'src/app/actions'
import { SuperchargeButtonType } from 'src/app/types'
import { REMOTE_CONFIG_VALUES_DEFAULTS } from 'src/firebase/remoteConfigValuesDefaults'
import { Screens } from 'src/navigator/Screens'
import { getRehydratePayload, REHYDRATE, RehydrateAction } from 'src/redux/persist-helper'

export interface State {
  loggedIn: boolean
  numberVerified: boolean
  analyticsEnabled: boolean
  requirePinOnAppOpen: boolean
  appState: AppState
  locked: boolean
  lastTimeBackgrounded: number
  sessionId: string
  minVersion: string | null
  celoEducationUri: string | null
  celoEuroEnabled: boolean
  dappListApiUrl: string | null
  inviteModalVisible: boolean
  activeScreen: Screens
  hideVerification: boolean
  showRaiseDailyLimitTarget: string | undefined
  walletConnectV1Enabled: boolean
  walletConnectV2Enabled: boolean
  rewardsPercent: number
  rewardsStartDate: number
  rewardsMax: number
  rewardsMin: number
  rewardsABTestThreshold: string
  // In 1.13 we had a critical error which requires a migration to fix. See |verificationMigration.ts|
  // for the migration code. We can remove all the code associated with this after some time has passed.
  ranVerificationMigrationAt: number | null | undefined
  logPhoneNumberTypeEnabled: boolean
  googleMobileServicesAvailable: boolean | undefined
  huaweiMobileServicesAvailable: boolean | undefined
  pincodeUseExpandedBlocklist: boolean
  rewardPillText?: {
    [lang: string]: string
  }
  cashInButtonExpEnabled: boolean
  rampCashInButtonExpEnabled: boolean
  multiTokenShowHomeBalances: boolean
  multiTokenUseSendFlow: boolean
  multiTokenUseUpdatedFeed: boolean
  linkBankAccountEnabled: boolean
  sentryTracesSampleRate: number
  sentryNetworkErrors: string[]
  superchargeButtonType: SuperchargeButtonType
}

const initialState = {
  loading: false,
  loggedIn: false,
  numberVerified: false,
  analyticsEnabled: true,
  requirePinOnAppOpen: false,
  appState: AppState.Active,
  locked: false,
  lastTimeBackgrounded: 0,
  sessionId: '',
  minVersion: null,
  celoEducationUri: null,
  celoEuroEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.celoEuroEnabled,
  dappListApiUrl: null,
  inviteModalVisible: false,
  activeScreen: Screens.Main,
  hideVerification: REMOTE_CONFIG_VALUES_DEFAULTS.hideVerification,
  showRaiseDailyLimitTarget: undefined,
  walletConnectV1Enabled: REMOTE_CONFIG_VALUES_DEFAULTS.walletConnectV1Enabled,
  walletConnectV2Enabled: REMOTE_CONFIG_VALUES_DEFAULTS.walletConnectV2Enabled,
  rewardsPercent: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsPercent,
  rewardsStartDate: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsStartDate,
  rewardsMax: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsMax,
  rewardsMin: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsMin,
  rewardsABTestThreshold: REMOTE_CONFIG_VALUES_DEFAULTS.rewardsABTestThreshold,
  ranVerificationMigrationAt: null,
  logPhoneNumberTypeEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.logPhoneNumberTypeEnabled,
  googleMobileServicesAvailable: undefined,
  huaweiMobileServicesAvailable: undefined,
  pincodeUseExpandedBlocklist: REMOTE_CONFIG_VALUES_DEFAULTS.pincodeUseExpandedBlocklist,
  rewardPillText: JSON.parse(REMOTE_CONFIG_VALUES_DEFAULTS.rewardPillText),
  cashInButtonExpEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.cashInButtonExpEnabled,
  rampCashInButtonExpEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.rampCashInButtonExpEnabled,
  multiTokenShowHomeBalances: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenShowHomeBalances,
  multiTokenUseSendFlow: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenUseSendFlow,
  multiTokenUseUpdatedFeed: REMOTE_CONFIG_VALUES_DEFAULTS.multiTokenUseUpdatedFeed,
  linkBankAccountEnabled: REMOTE_CONFIG_VALUES_DEFAULTS.linkBankAccountEnabled,
  sentryTracesSampleRate: REMOTE_CONFIG_VALUES_DEFAULTS.sentryTracesSampleRate,
  sentryNetworkErrors: REMOTE_CONFIG_VALUES_DEFAULTS.sentryNetworkErrors.split(','),
  superchargeButtonType: REMOTE_CONFIG_VALUES_DEFAULTS.superchargeButtonType,
}

export const appReducer = (
  state: State | undefined = initialState,
  action: ActionTypes | RehydrateAction
): State => {
  switch (action.type) {
    case REHYDRATE: {
      // Ignore some persisted properties
      const rehydratePayload = getRehydratePayload(action, 'app')
      return {
        ...state,
        ...rehydratePayload,
        appState: initialState.appState,
        locked: rehydratePayload.requirePinOnAppOpen ?? initialState.locked,
        sessionId: '',
      }
    }
    case Actions.SET_APP_STATE:
      let { appState, lastTimeBackgrounded } = state
      switch (action.state) {
        case 'background':
          if (Platform.OS === 'android') {
            lastTimeBackgrounded = Date.now()
          }
          appState = AppState.Background
          break
        case 'inactive': // occurs only on iOS
          lastTimeBackgrounded = Date.now()
          appState = AppState.Inactive
          break
        case 'active':
          appState = AppState.Active
          break
      }
      return {
        ...state,
        appState,
        lastTimeBackgrounded,
      }
    case Actions.SET_LOGGED_IN:
      return {
        ...state,
        loggedIn: action.loggedIn,
      }
    case Actions.SET_NUMBER_VERIFIED:
      return {
        ...state,
        numberVerified: action.numberVerified,
      }
    case Actions.RESET_APP_OPENED_STATE:
      return {
        ...state,
        loggedIn: false,
        numberVerified: false,
      }
    case Actions.SET_ANALYTICS_ENABLED:
      return {
        ...state,
        analyticsEnabled: action.enabled,
      }
    case Actions.SET_LOCK_WITH_PIN_ENABLED:
      return {
        ...state,
        requirePinOnAppOpen: action.enabled,
      }
    case Actions.LOCK:
      return {
        ...state,
        locked: true,
      }
    case Actions.UNLOCK:
      return {
        ...state,
        locked: false,
      }
    case Actions.SET_SESSION_ID:
      return {
        ...state,
        sessionId: action.sessionId,
      }
    case Actions.MIN_APP_VERSION_DETERMINED:
      return {
        ...state,
        minVersion: action.minVersion,
      }
    case Actions.UPDATE_REMOTE_CONFIG_VALUES:
      return {
        ...state,
        hideVerification: action.configValues.hideVerification,
        showRaiseDailyLimitTarget: action.configValues.showRaiseDailyLimitTarget,
        celoEducationUri: action.configValues.celoEducationUri,
        celoEuroEnabled: action.configValues.celoEuroEnabled,
        dappListApiUrl: action.configValues.dappListApiUrl,
        walletConnectV1Enabled: action.configValues.walletConnectV1Enabled,
        walletConnectV2Enabled: action.configValues.walletConnectV2Enabled,
        rewardsPercent: action.configValues.rewardsPercent,
        rewardsStartDate: action.configValues.rewardsStartDate,
        rewardsMax: action.configValues.rewardsMax,
        rewardsMin: action.configValues.rewardsMin,
        rewardsABTestThreshold: action.configValues.rewardsABTestThreshold,
        logPhoneNumberTypeEnabled: action.configValues.logPhoneNumberTypeEnabled,
        pincodeUseExpandedBlocklist: action.configValues.pincodeUseExpandedBlocklist,
        rewardPillText: JSON.parse(action.configValues.rewardPillText),
        cashInButtonExpEnabled: action.configValues.cashInButtonExpEnabled,
        rampCashInButtonExpEnabled: action.configValues.rampCashInButtonExpEnabled,
        multiTokenShowHomeBalances: action.configValues.multiTokenShowHomeBalances,
        multiTokenUseSendFlow: action.configValues.multiTokenUseSendFlow,
        multiTokenUseUpdatedFeed: action.configValues.multiTokenUseUpdatedFeed,
        linkBankAccountEnabled: action.configValues.linkBankAccountEnabled,
        sentryTracesSampleRate: action.configValues.sentryTracesSampleRate,
        sentryNetworkErrors: action.configValues.sentryNetworkErrors,
        superchargeButtonType: action.configValues.superchargeButtonType,
      }
    case Actions.TOGGLE_INVITE_MODAL:
      return {
        ...state,
        inviteModalVisible: action.inviteModalVisible,
      }
    case Actions.ACTIVE_SCREEN_CHANGED:
      return {
        ...state,
        activeScreen: action.activeScreen,
      }
    case Actions.VERIFICATION_MIGRATION_RAN:
      return {
        ...state,
        ranVerificationMigrationAt: action.now,
        numberVerified: action.isVerified,
      }
    case Actions.ANDROID_MOBILE_SERVICES_AVAILABILITY_CHECKED:
      return {
        ...state,
        googleMobileServicesAvailable: action.googleIsAvailable,
        huaweiMobileServicesAvailable: action.huaweiIsAvailable,
      }
    default:
      return state
  }
}
