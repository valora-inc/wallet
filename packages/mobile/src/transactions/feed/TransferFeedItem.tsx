import Touchable from '@celo/react-components/components/Touchable'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import variables from '@celo/react-components/styles/variables'
import BigNumber from 'bignumber.js'
import React from 'react'
import { StyleSheet, Text, View } from 'react-native'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import ContactCircle from 'src/components/ContactCircle'
import TokenDisplay from 'src/components/TokenDisplay'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { useTransferFeedDetails } from 'src/transactions/transferFeedUtils'
import { TokenTransfer } from 'src/transactions/types'

const AVATAR_SIZE = 40

interface Props {
  transfer: TokenTransfer
}

function TransferFeedItem({ transfer }: Props) {
  const { amount } = transfer

  const openTransferDetails = () => {
    // TODO: Allow opening detail screen.
    // navigateToPaymentTransferReview(type, timestamp, {
    //   comment: getDecryptedTransferFeedComment(comment ?? null, commentKey, type),
    //   amount,
    //   recipient,
    //   type,
    //   // fee TODO: add fee here.
    // })
    ValoraAnalytics.track(HomeEvents.transaction_feed_item_select)
  }

  const { title, subtitle, recipient } = useTransferFeedDetails(transfer)
  const colorStyle = new BigNumber(amount.value).isPositive() ? { color: colors.greenUI } : {}

  return (
    <Touchable disabled={false} onPress={openTransferDetails}>
      <View style={styles.container}>
        <View style={styles.iconContainer}>
          {<ContactCircle recipient={recipient} size={AVATAR_SIZE} />}
        </View>
        <View style={styles.contentContainer}>
          <View style={styles.titleContainer}>
            <Text style={styles.title} testID={'TransferFeedItem/title'}>
              {title}
            </Text>
            <TokenDisplay
              amount={amount.value}
              tokenAddress={amount.tokenAddress}
              currencyInfo={
                amount.localAmount
                  ? {
                      localCurrencyCode: amount.localAmount.currencyCode as LocalCurrencyCode,
                      localExchangeRate: amount.localAmount.exchangeRate,
                    }
                  : undefined
              }
              showExplicitPositiveSign={true}
              style={[styles.amount, colorStyle]}
              testID={'TransferFeedItem/amount'}
            />
          </View>
          <View style={styles.titleContainer}>
            <Text style={styles.subtitle} testID={'TransferFeedItem/subtitle'}>
              {subtitle}
            </Text>
            <TokenDisplay
              amount={amount.value}
              tokenAddress={amount.tokenAddress}
              showLocalAmount={false}
              showSymbol={true}
              hideSign={true}
              style={styles.tokenAmount}
              testID={'TransferFeedItem/tokenAmount'}
            />
          </View>
        </View>
      </View>
    </Touchable>
  )
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    flex: 1,
    paddingVertical: 12,
    paddingHorizontal: variables.contentPadding,
  },
  iconContainer: {
    flexDirection: 'column',
    alignItems: 'center',
  },
  contentContainer: {
    flex: 1,
    marginLeft: variables.contentPadding,
  },
  titleContainer: {
    flexDirection: 'row',
    marginTop: -1,
  },
  title: {
    ...fontStyles.regular500,
    flexShrink: 1,
  },
  subtitle: {
    ...fontStyles.small,
    color: colors.gray4,
    paddingTop: 2,
  },
  amount: {
    ...fontStyles.regular500,
    marginLeft: 'auto',
    paddingLeft: 10,
    minWidth: '40%',
    textAlign: 'right',
    flexWrap: 'wrap',
  },
  tokenAmount: {
    ...fontStyles.small,
    color: colors.gray4,
    paddingTop: 2,
    paddingLeft: 10,
    marginLeft: 'auto',
    minWidth: '40%',
    textAlign: 'right',
    flexWrap: 'wrap',
  },
})

export default TransferFeedItem
