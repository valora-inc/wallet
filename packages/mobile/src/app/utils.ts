import { DappRequestOrigin } from 'src/analytics/types'
import { ActiveDapp } from 'src/app/reducers'

// Assume that if we have an active dapp, any WC or DappKit request comes from the in-app webview.
// Note that this may be incorrect if the user interacts with another dapp while the dapp webview is open.
// I thought about comparing the request url to the active dapp url, but that could fail too if the dapp url we have is different.
// For instance the dapp-list could contain a "example.com", while the request is "dapp.something.com".
export function getDappRequestOrigin(activeDapp: ActiveDapp | null) {
  if (activeDapp) {
    return DappRequestOrigin.InAppWebView
  }

  return DappRequestOrigin.External
}
