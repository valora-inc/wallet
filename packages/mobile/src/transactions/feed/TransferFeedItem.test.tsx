import { render } from '@testing-library/react-native'
import React from 'react'
import { Provider } from 'react-redux'
import { ReactTestInstance } from 'react-test-renderer'
import { formatShortenedAddress } from 'src/components/ShortenedAddress'
import { LocalCurrencyCode } from 'src/localCurrency/consts'
import { RootState } from 'src/redux/reducers'
import TransferFeedItem from 'src/transactions/feed/TransferFeedItem'
import {
  Fee,
  TokenAmount,
  TokenTransactionTypeV2,
  TokenTransferMetadata,
  TransactionStatus,
} from 'src/transactions/types'
import { createMockStore, getElementText, RecursivePartial } from 'test/utils'
import { mockCeloAddress, mockCusdAddress, mockName, mockTestTokenAddress } from 'test/values'

const MOCK_TX_HASH = '0x006b866d20452a24d1d90c7514422188cc7c5d873e2f1ed661ec3f810ad5331c'
const MOCK_ADDRESS = '0xFdd8bD58115FfBf04e47411c1d228eCC45E93075'
const MOCK_E164_NUMBER = '+14155550001'
const MOCK_CONTACT = {
  name: mockName,
  displayNumber: '14155550001',
  e164PhoneNumber: MOCK_E164_NUMBER,
  contactId: 'contactId',
  address: MOCK_ADDRESS,
}

describe('TransferFeedItem', () => {
  function renderScreen({
    storeOverrides = {},
    type = TokenTransactionTypeV2.Sent,
    amount = {
      tokenAddress: mockCusdAddress,
      value: 10,
    },
    metadata = {},
    fees = [],
  }: {
    type?: TokenTransactionTypeV2
    amount?: TokenAmount
    metadata?: TokenTransferMetadata
    fees?: Fee[]
    storeOverrides?: RecursivePartial<RootState>
  }) {
    const store = createMockStore({
      ...storeOverrides,
    })

    const tree = render(
      <Provider store={store}>
        <TransferFeedItem
          transfer={{
            __typename: 'TokenTransferV2',
            type,
            status: TransactionStatus.Complete,
            transactionHash: MOCK_TX_HASH,
            timestamp: 1234,
            block: '2345',
            address: MOCK_ADDRESS,
            amount,
            metadata,
            fees,
          }}
        />
      </Provider>
    )

    return {
      store,
      ...tree,
    }
  }

  it('renders correctly', async () => {
    const tree = renderScreen({})
    expect(tree).toMatchSnapshot()
  })

  function expectDisplay({
    getByTestId,
    queryByTestId,
    expectedTitleSections,
    expectedSubtitleSections,
    expectedAmount,
    expectedTokenAmount,
  }: {
    getByTestId: (testId: string) => ReactTestInstance
    queryByTestId?: (testId: string) => ReactTestInstance | null
    expectedTitleSections: string[]
    expectedSubtitleSections: string[]
    expectedAmount: string
    expectedTokenAmount: string | null
  }) {
    const title = getElementText(getByTestId('TransferFeedItem/title'))
    for (const titleSection of expectedTitleSections) {
      expect(title).toContain(titleSection)
    }

    const subtitle = getElementText(getByTestId('TransferFeedItem/subtitle'))
    for (const subtitleSection of expectedSubtitleSections) {
      expect(subtitle).toContain(subtitleSection)
    }

    const amountDisplay = getByTestId('TransferFeedItem/amount')
    expect(getElementText(amountDisplay)).toEqual(expectedAmount)

    if (expectedTokenAmount) {
      const tokenDisplay = getByTestId('TransferFeedItem/tokenAmount')
      expect(getElementText(tokenDisplay)).toEqual(expectedTokenAmount)
    } else {
      expect(queryByTestId!('TransferFeedItem/tokenAmount')).toBeNull()
    }
  }

  it('renders correctly for outgoing transfers to unknown address', async () => {
    const { getByTestId } = renderScreen({})
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', formatShortenedAddress(MOCK_ADDRESS)],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for incoming transfers from unknown address', async () => {
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.Received,
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemReceivedTitle', formatShortenedAddress(MOCK_ADDRESS)],
      expectedSubtitleSections: ['feedItemReceivedInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers to phone contact', async () => {
    const { getByTestId } = renderScreen({
      storeOverrides: {
        identity: { addressToE164Number: { [MOCK_ADDRESS]: MOCK_E164_NUMBER } },
        recipients: {
          phoneRecipientCache: {
            [MOCK_E164_NUMBER]: MOCK_CONTACT,
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', mockName],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers to recent contact', async () => {
    const { getByTestId } = renderScreen({
      storeOverrides: {
        identity: { addressToE164Number: { [MOCK_ADDRESS]: MOCK_E164_NUMBER } },
        transactions: {
          recentTxRecipientsCache: {
            [MOCK_E164_NUMBER]: MOCK_CONTACT,
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', mockName],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers to phone number', async () => {
    const { getByTestId } = renderScreen({
      storeOverrides: {
        identity: { addressToE164Number: { [MOCK_ADDRESS]: MOCK_E164_NUMBER } },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', MOCK_E164_NUMBER],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers to Valora recipient', async () => {
    const { getByTestId } = renderScreen({
      storeOverrides: {
        recipients: {
          valoraRecipientCache: {
            [MOCK_ADDRESS]: { address: MOCK_ADDRESS, name: mockName },
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', mockName],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers with default title/subtitle', async () => {
    const { getByTestId } = renderScreen({
      metadata: {
        title: 'a title',
        subtitle: 'a subtitle',
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', 'a title'],
      expectedSubtitleSections: ['feedItemSentInfo', 'a subtitle'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  // TODO: Also test with encrypted comment.
  it('renders correctly for transfers with comments', async () => {
    const { getByTestId } = renderScreen({
      metadata: {
        title: 'a title',
        subtitle: 'a subtitle',
        comment: 'Hello World',
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', 'a title'],
      expectedSubtitleSections: ['feedItemSentInfo', 'Hello World'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers to a known provider', async () => {
    const { getByTestId } = renderScreen({
      storeOverrides: {
        fiatExchanges: {
          txHashToProvider: {
            [MOCK_TX_HASH]: {
              name: 'Simplex',
              icon: 'hi',
            },
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', 'Simplex'],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers from CELO rewards', async () => {
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.Received,
      storeOverrides: {
        identity: {
          addressToDisplayName: {
            [MOCK_ADDRESS]: {
              name: 'CELO Rewards',
              imageUrl: 'hi',
              isCeloRewardSender: true,
            },
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemCeloRewardReceivedTitle'],
      expectedSubtitleSections: ['feedItemRewardReceivedInfo'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers from Supercharge rewards', async () => {
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.Received,
      storeOverrides: {
        recipients: {
          rewardsSenders: [MOCK_ADDRESS],
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemRewardReceivedTitle'],
      expectedSubtitleSections: ['feedItemRewardReceivedInfo'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers from invite rewards', async () => {
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.Received,
      storeOverrides: {
        recipients: {
          inviteRewardsSenders: [MOCK_ADDRESS],
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemInviteRewardReceivedTitle'],
      expectedSubtitleSections: ['feedItemInviteRewardReceivedInfo'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for sent invites', async () => {
    const contactPhoneNumber = '+14155553695'
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.InviteSent,
      storeOverrides: {
        transactions: {
          inviteTransactions: {
            [MOCK_TX_HASH]: {
              recipientIdentifier:
                '0xd2c326a68d07b060be189bbe05a9abee8f10573a23cd448122d2efd122b4e05a',
              paymentId: '0x72a8898b40bD3CAf7e63664b61dBBa2E98fC123D',
            },
          },
        },
        recipients: {
          phoneRecipientCache: {
            [contactPhoneNumber]: {
              e164PhoneNumber: contactPhoneNumber,
              name: 'Kate Bell',
            },
          },
        },
        identity: {
          e164NumberToSalt: {
            [contactPhoneNumber]: 'hmq1hMZ08BYOg',
          },
        },
      },
    })

    expectDisplay({
      getByTestId,
      expectedTitleSections: [
        'feedItemEscrowSentTitle, {"context":null,"nameOrNumber":"Kate Bell"}',
      ],
      expectedSubtitleSections: ['feedItemEscrowSentInfo, {"context":"noComment"}'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for transfers from a known provider', async () => {
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.Received,
      storeOverrides: {
        fiatExchanges: {
          txHashToProvider: {
            [MOCK_TX_HASH]: {
              name: 'Simplex',
              icon: 'hi',
            },
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemReceivedTitle', 'Simplex'],
      expectedSubtitleSections: ['tokenDeposit', 'cUSD'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  // TODO: Also test with encrypted comment.
  it('renders correctly for transfers received with comments', async () => {
    const { getByTestId } = renderScreen({
      type: TokenTransactionTypeV2.Received,
      metadata: {
        title: 'a title',
        subtitle: 'a subtitle',
        comment: 'Hello World',
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemReceivedTitle', 'a title'],
      expectedSubtitleSections: ['feedItemReceivedInfo', 'Hello World'],
      expectedAmount: '+₱13.30',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly with custom local exchange rate', async () => {
    const { getByTestId } = renderScreen({
      amount: {
        tokenAddress: mockCusdAddress,
        value: 10,
        localAmount: {
          currencyCode: LocalCurrencyCode.PHP,
          exchangeRate: '1.5',
          value: '15',
        },
      },
    })
    expectDisplay({
      getByTestId,
      expectedTitleSections: ['feedItemSentTitle', formatShortenedAddress(MOCK_ADDRESS)],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+₱15.00',
      expectedTokenAmount: '10.00 cUSD',
    })
  })

  it('renders correctly for tokens without usd price', async () => {
    const { getByTestId, queryByTestId } = renderScreen({
      amount: {
        tokenAddress: mockTestTokenAddress,
        value: 10,
      },
      storeOverrides: {
        tokens: {
          tokenBalances: {
            [mockTestTokenAddress]: {
              address: mockTestTokenAddress,
              symbol: 'TT',
              balance: '50',
            },
          },
        },
      },
    })
    expectDisplay({
      getByTestId,
      queryByTestId,
      expectedTitleSections: ['feedItemSentTitle', formatShortenedAddress(MOCK_ADDRESS)],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+10.00 TT',
      expectedTokenAmount: null,
    })
  })

  it('renders the localAmount correctly when set', async () => {
    const { getByTestId, queryByTestId } = renderScreen({
      amount: {
        tokenAddress: mockCeloAddress,
        value: 10,
        localAmount: {
          currencyCode: 'EUR',
          exchangeRate: '0.4',
          value: '4',
        },
      },
    })
    expectDisplay({
      getByTestId,
      queryByTestId,
      expectedTitleSections: ['feedItemSentTitle', formatShortenedAddress(MOCK_ADDRESS)],
      expectedSubtitleSections: ['feedItemSentInfo', 'noComment'],
      expectedAmount: '+€4.00',
      expectedTokenAmount: '10.00 CELO',
    })
  })
})
