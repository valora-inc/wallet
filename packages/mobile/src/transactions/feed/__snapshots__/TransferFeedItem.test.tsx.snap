// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransferFeedItem renders correctly 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  nativeBackgroundAndroid={
    Object {
      "attribute": "selectableItemBackground",
      "rippleRadius": undefined,
      "type": "ThemeAttrAndroid",
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 16,
        "paddingVertical": 12,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
              },
              Object {
                "backgroundColor": "hsl(101, 53%, 93%)",
                "borderRadius": 20,
                "height": 40,
                "width": 40,
              },
            ]
          }
        >
          <svg
            fill="hsl(101, 53%, 93%)"
            height="40"
            style={Object {}}
            viewBox="0 0 40 40"
            width="40"
          />
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginLeft": 16,
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": -1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flexShrink": 1,
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
          testID="TransferFeedItem/title"
        >
          feedItemSentTitle, {"displayName":"feedItemAddress, {\\"address\\":\\"0xFdd8...3075\\"}"}
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "flexWrap": "wrap",
                "fontFamily": "Inter-Medium",
                "fontSize": 16,
                "lineHeight": 22,
                "marginLeft": "auto",
                "minWidth": "40%",
                "paddingLeft": 10,
                "textAlign": "right",
              },
              Object {
                "color": "#1AB775",
              },
            ]
          }
          testID="TransferFeedItem/amount"
        >
          +
          â‚±
          13.30
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "marginTop": -1,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 14,
              "lineHeight": 18,
              "paddingTop": 2,
            }
          }
          testID="TransferFeedItem/subtitle"
        >
          feedItemSentInfo, {"context":"noComment"}
        </Text>
        <Text
          style={
            Object {
              "color": "#9CA4A9",
              "flexWrap": "wrap",
              "fontFamily": "Inter-Regular",
              "fontSize": 14,
              "lineHeight": 18,
              "marginLeft": "auto",
              "minWidth": "40%",
              "paddingLeft": 10,
              "paddingTop": 2,
              "textAlign": "right",
            }
          }
          testID="TransferFeedItem/tokenAmount"
        >
          
          10.00
           cUSD
        </Text>
      </View>
    </View>
  </View>
</View>
`;
